src/software/amazon/ion/Timestamp.java

```diff
@@ -690,6 +699,28 @@ private Timestamp(BigDecimal millis, Integer localOffset)
     {
         if (millis == null) throw new NullPointerException("millis is null");
 
+        // check bounds to avoid hanging when calling longValue() on decimals with large positive exponents,
+        // e.g. 1e10000000
+        if(millis.compareTo(MINIMUM_ALLOWED_FRACTIONAL_MILLIS) < 0 ||
+            millis.compareTo(MAXIMUM_ALLOWED_FRACTIONAL_MILLIS) > 0) {
+            throw new IllegalArgumentException("millis: " + millis + " is outside of valid the range: from "
+                + MINIMUM_ALLOWED_FRACTIONAL_MILLIS
+                + " to "
+                + MAXIMUM_ALLOWED_FRACTIONAL_MILLIS
+                + " both inclusive");
+        }
+
+        // check scale size to avoid hanging on methods that need to inflate the fractional bigDecimal
+        // Examples: toString and toMillis
+        if(Math.abs(millis.scale()) >= MAXIMUM_ALLOWED_MILLIS_SCALE) {
```

Comment:
As noted previously, this violates the library design goals. We don't set bounds on the size of strings, or on decimals or ints, or anything else, and we shouldn't do it here. This is rejecting perfectly valid Ion data, and I don't find that to be acceptable (default) behavior for the reference implementation of Ion.

If the intent is to protect other operations from having undesirable performance, then that should be addressed at that point.  In the case of converting to millis, that is not hard to do.

It would be reasonable to have a configuration point that allows applications to reign in the bounds of some things, but it shouldn't be done ad-hoc nor by default.